Mappings: Fix not_analyzed string fields to error when position_increment_gap is set

rjernst opened the pull request:
Currently if a string field is not_analyzed, but a
position_increment_gap is set, it will lookup the default analyzer and
set it, along with the position_increment_gap, before the code which
handles setting the keyword analyzer for not_analyzed fields has a
chance to run. This change adds a parsing check and test for that case.

rmuir commented:
Analyzer has two gaps: one for positions, and another for offsets.

Confusingly, ES previously named the positions one `positions_offset_gap` (see https://github.com/elastic/elasticsearch/issues/12562), but i've seen separately an `offsetGap` (hopefully its not configurable, its basically going to guarantee broken highlighting...)

Are both the bases covered?

rjernst commented:
@rmuir Yes, we have tests that ensure position_offset_gap is rejected.

rmuir commented:
i'm not really concerned about `position_offset_gap`, which was simply a confusing name that ES mapped with Analyzer's position increment gap.

Lucene has two gaps, and looking at the source code ES configures both of them (https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzer.java#L47)

So I want to make sure its prevented in both cases.

rjernst commented:
Mappings do not have a way to set the actual lucene offset gap.

rmuir commented:
OK, then this change is fine. Separately, exposing that offset gap in any way should be seriously reconsidered: from a practical perspective, all it will accomplish is to break highlighting.

rmuir commented:
Also i spoke with @rjernst, a more general solution would be this check:
```
// position_increment_gap is senseless unless positions are indexed
if (indexOptions.compareTo(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS) < 0) {
  error(...)
}
```

rjernst commented:
@rmuir I updated to use the check, added another test for disabling positions, and also added this for `text` fields which I realized were prone to this if positions were disabled.

jpountz commented:
LGTM

rjernst commented:
2.4: f91e605
